### PR TITLE
Track terminating pods

### PR DESCRIPTION
**Pull request by joshimoo**
We changed the behavior so that we only set lastPodRefAt once there are
no more pods present, previously this was already set when the last pod
went into terminating state.

I added a unit test for both cases, to verify the behavior of tracking in
terminating state and lastPodRefAt setting once there are no more pods
present for that volume.

longhorn/longhorn#2874

Signed-off-by: Joshua Moody <joshua.moody@suse.com>

### CODE REVIEW

**Comment by joshimoo** [comment]
Testing https://github.com/longhorn/longhorn/issues/2960 to see if this fixes that issue, will have to make some additional changes.

**Review by shuo-wu — APPROVED** [review]

**Comment by joshimoo** [comment]
Added a fix for https://github.com/longhorn/longhorn/issues/2960
by adding a stopping state, the controller will transition the share-manager into stopping state, when the volume is no longer beeing used, it will then wait for the share-manager pod to be cleaned up before transitioning into stopped state.

This breaks the previous loop of:
I.E a single pod workload, we could end up in a bad loop of
starting (new workload) -> error (remount) -> stopped (cleanup sm)

Since we would transition from stopped into starting, but the old pod has been marked for deletion but not cleaned up yet. This would signal error state, so we end up in the above cycle.

**Comment by joshimoo** [comment]
/hold merge
Review please, will merge this after finishing my final test run.

File: controller/share_manager_controller.go
Comment (shuo-wu) [inline controller/share_manager_controller.go:443]: NIT: If `sm.Status.State != types.ShareManagerStateStopping && sm.Status.State != types.ShareManagerStateStopped` is used here, there is no need to re-check it below when setting `isNotNeeded` to `true`. Or vice versa.

Furthermore, I am thinking if we really need to check the share manager state before detachment. Is this check for avoiding detaching a stopped volume?

File: controller/share_manager_controller.go
Comment (shuo-wu) [inline controller/share_manager_controller.go:447]: NIT: print out the error log

File: controller/share_manager_controller.go
Comment (joshimoo) [reply controller/share_manager_controller.go:443]: Correct, this is so that if the share-manager is stopped we don't mess with the volume.
The extra state check is a little redundant but it helps make it more clear in this case.

File: controller/share_manager_controller.go
Comment (joshimoo) [reply controller/share_manager_controller.go:447]: The error log is already handled in the detachShareManagerVolume.
The only other error that is possible is conflict, which we don't need to print.

**Comment by joshimoo** [comment]
@shuo-wu @innobead my testing looks good can we merge this?

**Review by shuo-wu — APPROVED** [review]

**Comment by innobead** [comment]
LGTM

**Pull request merged**
